clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1231,7 +1234,13 @@ public void assign(Collection<TopicPartition> partitions) {
                         client.pollNoWakeup();
                     }
 
-                    return this.interceptors.onConsume(new ConsumerRecords<>(records));
+                    // after the long poll, we should filter the returned data if their belonging
+                    // partitions are no longer owned by the consumer
+                    final Set<TopicPartition> assignedPartitions = subscriptions.assignedPartitions();
+                    return this.interceptors.onConsume(new ConsumerRecords<>(records.entrySet()
```

Comment:
A meta-comment here.

I think that filtering the records under most conditions would be redundant since the assignment usually does not change. Since the assignment remains static unless a rebalance occurs, I think that  we only need to filter records when we _have found that the assignment has changed_ like after a rebalance is finished. In the case where  the assignment does not change, this segment of the code has _no effect_ since the assignment has remained the same. It probably would act as a substantial performance hit since this portion of the code is often heavily congested.

We probably should have some sort of check that only filters the records when the assignment has changed.